### PR TITLE
[WIP] Refactor Base58 parsing, nullabify more

### PR DESCRIPTION
**Pull request by NicolasDorier**

### CODE REVIEW

File: NBitcoin/JsonConverters/CoinJsonConverter.cs [hunk 1]
```diff
@@ -62,7 +62,7 @@ public ICoin ToCoin(string path)
 
 				
 				var coin = RedeemScript == null ? new Coin(new OutPoint(TransactionId, index), new TxOut(Value, ScriptPubKey)) : new ScriptCoin(new OutPoint(TransactionId, index), new TxOut(Value, ScriptPubKey), RedeemScript);
-				if (AssetId != null)
+				if (!(AssetId is null))
```
Comment (yahiheb) [inline NBitcoin/JsonConverters/CoinJsonConverter.cs]: In C# 8.0 there is this `if (variable is { })` which reads as if variable is something or is object (not null)
OR this `if (variable is object)`
The latter is maybe more readable but my preference is the former as there are more things you can do with it,. Check this video.
https://www.youtube.com/watch?v=TJiLhRPgyq4&t=622s
https://intellitect.com/check-for-null-not-null/

File: NBitcoin/Base58Data.cs [hunk 2]
```diff
@@ -34,67 +35,50 @@ public Network Network
 			}
 		}
 
-		protected void Init<T>(string base64, Network expectedNetwork = null) where T : Base58Data
+		protected Base58Data(string base58, Network expectedNetwork)
 		{
+			if (base58 == null)
```
Comment (yahiheb) [inline NBitcoin/Base58Data.cs]:
```suggestion
			if (base58 is null)
```

File: NBitcoin/Base58Data.cs [hunk 3]
```diff
@@ -34,67 +35,50 @@ public Network Network
 			}
 		}
 
-		protected void Init<T>(string base64, Network expectedNetwork = null) where T : Base58Data
+		protected Base58Data(string base58, Network expectedNetwork)
 		{
+			if (base58 == null)
+				throw new ArgumentNullException(nameof(base58));
+			if (expectedNetwork == null)
```
Comment (yahiheb) [inline NBitcoin/Base58Data.cs]:
```suggestion
			if (expectedNetwork is null)
```

File: NBitcoin/Base58Data.cs [hunk 4]
```diff
@@ -34,67 +35,50 @@ public Network Network
...
 			}
 
 			if (!IsValid)
 				throw new FormatException("Invalid " + this.GetType().Name);
 
 		}
 
+		protected Base58Data(byte[] rawBytes, Network network)
+		{
+			if (network == null)
```
Comment (yahiheb) [inline NBitcoin/Base58Data.cs]:
```suggestion
			if (network is null)
```

File: NBitcoin/JsonConverters/CoinJsonConverter.cs (hunk 1 above)
Comment (NicolasDorier) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs]: WTF how could I be unaware of `if(variable is { })` ?!?

Comment (NicolasDorier) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs]: `variable is { }` is what `variable is var` should have been from day one.

Comment (nopara73) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs]: @NicolasDorier Look at the youtube video @yahiheb linked. There are crazy things to do with `is { }`, we decided on using that from yesterday.

Comment (NicolasDorier) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs]: I asked them about it on https://twitter.com/NicolasDorier/status/1196449885538549761?s=20

Apparently `{}` is not a keyword but part of the "property based pattern matching". Which I completely ignored when I read as I did not see the purpose.

Comment (yahiheb) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs:65]: @nopara73 @NicolasDorier https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#logical-patterns
With C# 9.0 you can simply write `if (AssetId is not null)`

Comment (NicolasDorier) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs:65]: @yahiheb oh my god they implemented it! I thought there was some tough discussion drama around it.

**Comment by NicolasDorier** [comment]
Closing this, tried to fix the PR on top of master, could not. This is too low priority to bother.

File: NBitcoin/JsonConverters/CoinJsonConverter.cs (hunk 1 above)
Comment (nopara73) [reply NBitcoin/JsonConverters/CoinJsonConverter.cs:65]: image

**Pull request closed without merging**